diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -852,10 +852,12 @@
 	lzma | uImage lzma
   KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x68737173
   IMAGE_SIZE := 14528k
-  IMAGES += recovery.bin
+  IMAGES += factory.bin recovery.bin
   IMAGE/recovery.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | \
 	append-rootfs | pad-rootfs | check-size | pad-to 14528k | \
 	append-loader-okli-uimage $(1) | pad-to 15616k
+  IMAGE/factory.bin := $$(IMAGE/recovery.bin) | dlink-sge-image | \
+	dlink-sge-signature COVR-P2500
 endef
 TARGET_DEVICES += dlink_covr-p2500-a1
 
